### ISSUE TITLE
The ASP.net SignalR webapp is unresponsive when requests volume is high. We suspect deadlock in it.

### ISSUE THREAD

**Issue by AlexChongMicrosoft** [SignalR/SignalR#4671]
### Expected  behavior
The webapp with SignalR should handle the requests well.

### Actual behavior
The webapp is unresponsive

### Steps to reproduce
It's an intermittent issue. We found when the current connections to the webapp are more than 2000, the webapp becomes unresponsive. We suspect the high volumne of the requests cause this issue.

In the hang dump, we found 2000+ threads are waiting a lock from 1 thread, but the thread is also waiting in System.Threading.Monitor.Wait. Per my insight, it's like a deadlock issue.

Please advise.

**Comment by AlexChongMicrosoft** [comment]
Is there someone still monitoring this project?

**Comment by AlexChongMicrosoft** [comment]
@BrennanConroy let me check the signalr logs following the doc you shared.

The call stack is below. If you wish, I can share the details via Teams. chuchong
ntdll!NtWaitForMultipleObjects+a    
KERNELBASE!WaitForMultipleObjectsEx+e1    
clr!WaitForMultipleObjectsEx_SO_TOLERANT+62    
clr!Thread::DoAppropriateWaitWorker+205    
clr!Thread::DoAppropriateWait+7d    
clr!CLREventBase::WaitEx+b6    
clr!Thread::Block+2c    
clr!SyncBlock::Wait+1c8    
clr!ObjectNative::WaitTimeout+e1    
mscorlib_ni!System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)+364    
mscorlib_ni!System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)+bb    
mscorlib_ni!System.Threading.Tasks.Task.InternalWait(Int32, System.Threading.CancellationToken)+1b1    
mscorlib_ni!System.Threading.Tasks.Task.Wait(Int32, System.Threading.CancellationToken)+39    
Microsoft.AspNet.SignalR.Messaging.ScaleoutStream.Send(System.Func`2<System.Object,System.Threading.Tasks.Task>, System.Object)+133    
Microsoft.AspNet.SignalR.Infrastructure.Connection.Send(Microsoft.AspNet.SignalR.ConnectionMessage)+172    
Microsoft.AspNet.SignalR.Transports.TransportConnectionExtensions.SendCommand(Microsoft.AspNet.SignalR.Transports.ITransportConnection, System.String, Microsoft.AspNet.SignalR.Messaging.CommandType)+94    
Microsoft.AspNet.SignalR.Transports.ForeverTransport+<>c__DisplayClass41_0.<ProcessReceiveRequest>b__0()+1e    
Microsoft.AspNet.SignalR.Transports.ForeverTransport+<>c__DisplayClass41_0.<ProcessReceiveRequest>b__1()+16    
Microsoft.AspNet.SignalR.Transports.ForeverTransport.ProcessMessages(Microsoft.AspNet.SignalR.Transports.ITransportConnection, System.Func`1<System.Threading.Tasks.Task>)+2e1    
Microsoft.AspNet.SignalR.TaskAsyncHelper.FromMethod[[System.__Canon, mscorlib],[System.__Canon, mscorlib]](System.Func`3<System.__Canon,System.__Canon,System.Threading.Tasks.Task>, System.__Canon, System.__Canon)+21    
Microsoft.AspNet.SignalR.Transports.ForeverTransport.ProcessRequestCore(Microsoft.AspNet.SignalR.Transports.ITransportConnection)+12f    
Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler+<>c__DisplayClass5_0+<<RunWebSocketHandler>b__0>d.MoveNext()+84    
System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler+<>c__DisplayClass5_0+<<RunWebSocketHandler>b__0>d, Microsoft.AspNet.SignalR.Core]](<<RunWebSocketHandler>b__0>d ByRef)+80    
Microsoft.AspNet.SignalR.Owin.OwinWebSocketHandler+<>c__DisplayClass5_0.<RunWebSocketHandler>b__0()+66    
mscorlib_ni!System.Threading.Tasks.Task`1[[System.__Canon, mscorlib]].InnerInvoke()+4e    
mscorlib_ni!System.Threading.Tasks.Task.Execute()+47    
mscorlib_ni!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+172    
mscorlib_ni!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+15    
mscorlib_ni!System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)+231    
mscorlib_ni!System.Threading.Tasks.Task.ExecuteEntry(Boolean)+a1    
mscorlib_ni!System.Threading.ThreadPoolWorkQueue.Dispatch()+156    
clr!CallDescrWorkerInternal+83    
clr!CallDescrWorkerWithHandler+4e    
clr!MethodDescCallSite::CallTargetWorker+fa    
clr!QueueUserWorkItemManagedCallback+2a    
clr!ManagedThreadBase_DispatchInner+40    
clr!ManagedThreadBase_DispatchMiddle+6c    
clr!ManagedThreadBase_DispatchOuter+4c    
clr!ManagedThreadBase_DispatchInCorrectAD+15    
clr!Thread::DoADCallBack+26b    
clr!ManagedThreadBase_DispatchInner+2e57    
clr!ManagedThreadBase_DispatchMiddle+6c    
clr!ManagedThreadBase_DispatchOuter+4c    
clr!ManagedThreadBase_FullTransitionWithAD+2f    
clr!ManagedPerAppDomainTPCount::DispatchWorkItem+a4    
clr!ThreadpoolMgr::ExecuteWorkRequest+64    
clr!ThreadpoolMgr::WorkerThreadStart+f6    
clr!Thread::intermediateThreadProc+8b    
kernel32!BaseThreadInitThunk+22    
ntdll!RtlUserThreadStart+34

**Comment by AlexChongMicrosoft** [comment]
@BrennanConroy HI Brennan, how are you? Do you have any chance to check the above callstack? Meanwhile, I'm trying to ask the cx to enable the SignalR trace.

1 more question here, though ASP.NET SignalR is still supported, seems current main stream is ASP.net core SignalR and ASRS. May I know whether you will check and fix if ASP.net SignalR has something wrong?

**Comment by BrennanConroy** [comment]
> we found 2000+ threads are waiting a lock from 1 thread

Does that mean the 2000+ threads have Microsoft.AspNet.SignalR.Messaging.ScaleoutStream.Send at the top of their stacks? (Well Monitor.Wait and then ScaleoutStream.Send)

Are there any threads not waiting? There should be something trying to send to your scaleout provider. Is your scaleout provider really slow or overwhelmed?

**Comment by AlexChongMicrosoft** [comment]
@BrennanConroy I randomly checked some of the 2000+ threads, at the top of the stacks, yes, we can see Microsoft.AspNet.SignalR.Messaging.ScaleoutStream.Send as below.

~21
![image](https://user-images.githubusercontent.com/45751478/236134150-32ba066d-d9d0-442b-920f-d2c80a907b7e.png)

~57
![image](https://user-images.githubusercontent.com/45751478/236134232-f73f30d0-3543-4307-a72f-870dbcea2261.png)

~104
![image](https://user-images.githubusercontent.com/45751478/236134298-10e88a53-273e-42f4-a21c-938c959eda32.png)


Regarding the question "Are there any threads not waiting? There should be something trying to send to your scaleout provider. Is your scaleout provider really slow or overwhelmed?"
Yes, 22 and 23 are not waiting for the 24, but not sure they are asp.net signalR threads. Regarding the scaleout provider, the customer uses Redis as the backplane. I'm not sure whether it's slow or overwhelmed. I can ask such question. Do you have any suggestion to check the backplane?
![image](https://user-images.githubusercontent.com/45751478/236134678-6f65a4d7-58ac-4a34-ae52-b1d3778cb7d1.png)

**Comment by AlexChongMicrosoft** [comment]
@BrennanConroy I get the cx's feedback. The scaleout provider is currently an Azure hosted Redis instance, they have checked the metrics for it and it doesn't seem to be overloaded, it is sitting at around 20-30% load most of the time.  The user will attempt to scale this up to see if there is an effect on the SignalR issue.

**Comment by AlexChongMicrosoft** [comment]
@BrennanConroy, 
From the customer's feedback, when the SignalR app is unresponsive, the Redis backplane is around 5-10%load as below. Look forward to your feedback.

![image](https://github.com/SignalR/SignalR/assets/45751478/ae3c6da1-a982-4eba-aab1-a54b0a9edc25)

**Comment by AlexChongMicrosoft** [comment]
@BrennanConroy , how are you? This is a friendly follow up message.

I'm requesting the cx's SignalR scaleout trace following below doc. May I know is there something more you want to check?
Enable Tracing For SignalR Scaleout [https://learn.microsoft.com/en-us/aspnet/signalr/overview/performance/scaleout-in-signalr#enabling-tracing-for-signalr-scaleout](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Faspnet%2Fsignalr%2Foverview%2Fperformance%2Fscaleout-in-signalr%23enabling-tracing-for-signalr-scaleout&data=05%7C01%7CChuanyu.Chong%40microsoft.com%7Ca7d059a1733f49018b0608db55e66a19%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638198220452368462%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=aez21Lt8%2FPgyD67ucVQv0r%2BwpJ9tvSHSEjSvcM6YgXI%3D&reserved=0)

**Comment by eli-kara** [comment]
Hi,
We've encountered the same issue too, exactly this one, on a customer site. However, we are not using Redis scaleout but a different solution. The scaleout message bus was built in a similar fashion to the Redis one.

I believe there's a logical error in the code. Let's call the Send operation S and the stream open operation O. Consider this sequence of events:

S1, S2, S3, S4, S5, O, S6...
(simulating queueing before stream open).

The first 5 Sends will be queued into the TaskQueue. When you queue a new Task to the TaskQueue you pass it the "Send" function of the ScaleoutStream.
After the stream is opened, the queue's Drain is called. Now, if Send number 6 (e.g. S6) occurs before S1 to S5 managed to be scheduled, it will block on the drain Task, holding the lock object "_lockObj" of the ScaleoutStream.

The previous Send operations will block because they require the lock object too, but S6 has it locked and waiting on the tail of the queue to finish. Classic deadlock but masked due to the usage of lambda functions and the async nature of this code.

BTW, a solution would be to open the stream in a synchronized manner, as opposed to using an async Task to do it. Somebody probably decided opening a stream from the ctor is too dangerous so they used an async Task to do it (see RedisMessageBus).
We used the same approach in our backplane and have the same problem.

Ideally, you'd build the message bus, initialize it after object creation and register it.

**Comment by eli-kara** [comment]
BTW, to more easily reproduce this, you can put a sleep of 5 seconds before opening the stream. Example attached.
![RedisMessageBus_repro_deadlock](https://github.com/user-attachments/assets/f93e6149-e6d0-4ece-97c5-165c27efbb56)

**Comment by BrennanConroy** [comment]
> When you queue a new Task to the TaskQueue you pass it the "Send" function of the ScaleoutStream.
> Each needs to call ScaleoutStream::Send, which will now block since "_lockObj" is locked by the thread that handles S6

We pass it the non-locking `Send` function
https://github.com/SignalR/SignalR/blob/336c877743924c990aeefc75930d63e609e624ac/src/Microsoft.AspNet.SignalR.Core/Messaging/ScaleoutStream.cs#L171
so when the sends are being drained there isn't a lock being acquired.

However, I did notice that if the `RedisMessageBus` throws from its `Send` function in the case you described above, then we do acquire the lock and can result in a deadlock.
https://github.com/SignalR/SignalR/blob/336c877743924c990aeefc75930d63e609e624ac/src/Microsoft.AspNet.SignalR.Core/Messaging/ScaleoutStream.cs#L187
Is this what you are seeing? If not, we'll need more debugging details.

**Comment by eli-kara** [comment]
Hi, in our case we don't use the Redis scaleout, but a different backplane implemented similarly. We basically took the Kafka PoC code from Microsoft.AspNet.SignalR.Kafka and adjusted it to our needs..

However, I wanted to comment on your first comment.

The Enqueue method accepts a Func<object, Task> + another state object and the only Send function that matches that signature is the blocking version of Send. Hence I'm a bit confused as to why you claim the code sends the non-blocking version, which clearly has a different signature (receiving just an object). Maybe I'm missing something.

**Comment by BrennanConroy** [comment]
There are 2 `Send` functions in `ScaleoutStream`.
`public Task Send(Func<object, Task> send, object state)` which is equivalent to `Func<Func<object, Task>, object, Task>`.
`private static Task Send(object state)` which is equivalent to `Func<object, Task>`.

Like you said, the `Enqueue` method accepts a `Func<object, Task>` which only matches with the second `Send` method, which does not acquire the lock except in the exception case I mentioned in my previous comment.

So, the only deadlock I can see is from when the scaleout implementation throws during its `Send` method during one of S1-S5.

**Comment by eli-kara** [comment]
@BrennanConroy thanks, too many hours looking at the code. You're right of course. I looked again and found this stack trace in our dump file:
`
000000dd099fe850 00007ffa9d04b102 Microsoft.AspNet.SignalR.Messaging.ScaleoutStream.Open()
000000dd099fe8b0 00007ffa9d078eb3 Microsoft.AspNet.SignalR.Messaging.ScaleoutStreamManager.OnReceived(Int32, UInt64, Microsoft.AspNet.SignalR.Messaging.ScaleoutMessage)
000000dd099fe8f0 00007ffa9d0789a7 Microsoft.AspNet.SignalR.Kafka.KafkaMessageBus.OnMessage(Int32, UInt64, Byte[])
000000dd099fe970 00007ffa9d050e3c Microsoft.AspNet.SignalR.Kafka.KafkaMessageBus.Send(Int32, System.Collections.Generic.IList`1<Microsoft.AspNet.SignalR.Messaging.Message>)
000000dd099feaa0 00007ffa9d050988 Microsoft.AspNet.SignalR.Messaging.ScaleoutStream+SendContext.InvokeSend()
000000dd099feae0 00007ffa9d0507fd Microsoft.AspNet.SignalR.Messaging.ScaleoutStream.Send(System.Object)
000000dd099feb30 00007ffa9d05015a Microsoft.AspNet.SignalR.Infrastructure.TaskQueue.InvokeNext(System.Func`2<System.Object,System.Threading.Tasks.Task>, System.Object)
000000dd099feb70 00007ffa9d04fd00 Microsoft.AspNet.SignalR.TaskAsyncHelper+TaskRunners`2+<>c__DisplayClass2_0[[System.__Canon, mscorlib],[System.__Canon, mscorlib]].<RunTask>b__0(System.Threading.Tasks.Task)
000000dd099febc0 00007ffa9d04fa24 Microsoft.AspNet.SignalR.TaskAsyncHelper+<>c__71`1[[System.__Canon, mscorlib]].<RunWithPreservedCulture>b__71_0(System.Action`1<System.__Canon>, System.__Canon)
000000dd099febf0 00007ffa9d04f13d Microsoft.AspNet.SignalR.TaskAsyncHelper.RunWithPreservedCulture[[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.__Canon, mscorlib]](CulturePair, System.Func`3<System.__Canon,System.__Canon,System.__Canon>, System.__Canon, System.__Canon)
000000dd099fec60 00007ffa9d04ebbb Microsoft.AspNet.SignalR.TaskAsyncHelper.RunWithPreservedCulture[[System.__Canon, mscorlib]](CulturePair, System.Action`1<System.__Canon>, System.__Canon)
000000dd099fecf0 00007ffa9d04e2de Microsoft.AspNet.SignalR.TaskAsyncHelper+<>c__DisplayClass73_0.<ContinueWithPreservedCulture>b__0(System.Threading.Tasks.Task)
000000dd099fed30 00007ffaf792d0c7 System.Threading.Tasks.Task.Execute() [f:\dd\ndp\clr\src\BCL\system\threading\Tasks\Task.cs @ 2498]
`
The non-blocking Send is called alright, but it calls the bus's Send message (with the list of messages), and sure enough, in our implementation of the kafka message bus we call OnReceived (via OnMessage) which will go to ScaleoutStream.Open. If S6 reaches the Wait before the stack can finish OnReceived from the TaskQueue, we'll be blocked alright. That fully explains *our* deadlock.

This behavior differs from RedisMessageBus though. It doesn't call OnReceived during it's Send(int streamIndex, IList<Message> messages) function at all. The only way to have OnReceived called in such a way is if something came from the backplane.

@AlexChongMicrosoft since you have a dump file for your repro case, maybe give it another look. Perhaps something that involves the TaskQueue.
